applications/RomApplication/custom_strategies/rom_builder_and_solver.h

```diff
@@ -676,6 +410,347 @@ class ROMBuilderAndSolver : public BuilderAndSolver<TSparseSpace, TDenseSpace, T
+     * Computes the local contribution of an element or condition
+     */
+    template<typename TEntity>
+    std::tuple<LocalSystemMatrixType, LocalSystemVectorType> CalculateLocalContribution(
+        TEntity& rEntity,
+        AssemblyPrealocation& rPreAlloc,
+        TSchemeType& rScheme,
+        const ProcessInfo& rCurrentProcessInfo)
+    {
+        if (rEntity.IsDefined(ACTIVE) && rEntity.IsNot(ACTIVE))
+        {
+            rPreAlloc.romA = ZeroMatrix(mNumberOfRomModes, mNumberOfRomModes);
+            rPreAlloc.romB = ZeroVector(mNumberOfRomModes);
+            return std::tie(rPreAlloc.romA, rPreAlloc.romB);
+        }
+
+        // Calculate elemental contribution
+        rScheme.CalculateSystemContributions(rEntity, rPreAlloc.lhs, rPreAlloc.rhs, rPreAlloc.eq_id, rCurrentProcessInfo);
+        rEntity.GetDofList(rPreAlloc.dofs, rCurrentProcessInfo);
+
+        const SizeType ndofs = rPreAlloc.dofs.size();
+        ResizeIfNeeded(rPreAlloc.phiE, ndofs, mNumberOfRomModes);
+        ResizeIfNeeded(rPreAlloc.aux, ndofs, mNumberOfRomModes);
+
+        const auto &r_geom = rEntity.GetGeometry();
+        RomAuxiliaryUtilities::GetPhiElemental(rPreAlloc.phiE, rPreAlloc.dofs, r_geom, mMapPhi);
+
+        const double h_rom_weight = mHromSimulation ? rEntity.GetValue(HROM_WEIGHT) : 1.0;
+
+        noalias(rPreAlloc.aux) = prod(rPreAlloc.lhs, rPreAlloc.phiE);
+        noalias(rPreAlloc.romA) = prod(trans(rPreAlloc.phiE), rPreAlloc.aux) * h_rom_weight;
+        noalias(rPreAlloc.romB) = prod(trans(rPreAlloc.phiE), rPreAlloc.rhs) * h_rom_weight;
+
+        return std::tie(rPreAlloc.romA, rPreAlloc.romB);
+    }
+
+    /**
+     * Builds the reduced system of equations on rank 0 
+     */
+    void BuildROM(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart &rModelPart,
+        RomSystemMatrixType &A,
+        RomSystemVectorType &b)
+    {
+        KRATOS_TRY
+
+        // Define a dense matrix to hold the reduced problem
+        A = ZeroMatrix(mNumberOfRomModes, mNumberOfRomModes);
+        b = ZeroVector(mNumberOfRomModes);
+
+        // Build the system matrix by looping over elements and conditions and assembling to A
+        KRATOS_ERROR_IF(!pScheme) << "No scheme provided!" << std::endl;
+
+        // Get ProcessInfo from main model part
+        const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        // Assemble all entities
+        const auto assembling_timer = BuiltinTimer();
+
+        using SystemSumReducer = CombinedReduction<NonTrivialSumReduction<RomSystemMatrixType>, NonTrivialSumReduction<RomSystemVectorType>>;
+        AssemblyPrealocation prealloc(mNumberOfRomModes);
+
+        auto& elements = mHromSimulation ? mSelectedElements : rModelPart.Elements();
+        if(!elements.empty())
+        {
+            std::tie(A, b) =
+            block_for_each<SystemSumReducer>(elements, prealloc, 
+                [&](Element& r_element, AssemblyPrealocation& thread_prealloc)

```

Comment:
```suggestion
                [&](Element& r_element, AssemblyPrealocation& r_thread_prealloc)
```
Minor style suggestion.